index.d.ts

```diff
@@ -0,0 +1,233 @@
+	genre: string;
+	author: SongAuthor;
+	publishedAt: Date;
+	embedURL: string;
+	embed: Embed;
+	track: Optional<SongStreams>;
+	trackURL: string;
+	streamURL: string;
+	comments: Comment[];
+}
+
+export interface UserInfo extends BaseUser {
+	createdAt: Date;
+	avatarURL: string;
+	profile: string;
+	bannerURL: string;
+	followers: number;
+	following: number;
+	likesCount: number;
+	tracksCouint: number;
+	tracks: UserTracks[];
+	likes: UserLikes[];
+}
+
+export interface UserLikesAuthor {
+	name?: string;
+	username?: string;
+	profile?: string;
+}
+
+export interface UserLikes {
+	title: string;
+	url: string;
+	publishedAt: Date;
+	author: UserLikesAuthor;
+}
+
+export interface UserTracks {
+	title: string;
+	url: string;
+	publishedAt: Date;
+	genre: string;
+	author: string;
+	duration: number;
+}
+
+export interface SongInfoOptions {
+	fetchEmbed?: boolean;
+	fetchComments?: boolean;
+	fetchStreamURL?: boolean;
+	requestOptions?: RequestInit;
+}
+
+export class Client {
+	constructor(apiKey?: string, options?: ClientOptions);
+
+	options: ClientOptions;
+
+	apiVersion(force?: boolean): Promise<string>;
+	createAPIKey(key: string, fetch?: boolean): Promise<void>;
+	fetchStreamURL(trackURL: string): Promise<string>;
+	getEmbed(embedURL: string): Promise<Embed>;
+	getPlaylist(url: string, options?: PlaylistParseOptions): Promise<Playlist>;
+	getSongInfo(url: string, options?: SongInfoOptions): Promise<Song>;
+	getUser(username: string): Promise<UserInfo>;
+	search(query: string, type: SearchType): Promise<SearchResult[]>;
+}
+
+export class Downloader {
+	constructor();
+
+	downloadHLS(url: string, options?: m3u8Options): Promise<m3u8Stream>;
+	downloadProgressive(url: string, options?: RequestOptions): Promise<IncomingMessage>;
+}
+
+export class Embed {
+	constructor(data: object, embedURL: string);
+
+	author: EmbedAuthor;
+	description: string;
+	height: number;
+	provider: EmbedProvider;
+	thumbnailURL: string;
+	title: string;
+	type: string;
+	url: string;
+	version: number;
+	visualizer: string;
+	width: number;
+
+	toHTML(): string;
+	toJSON(): object;
+	toString(): string;
+}
+
+export class Song {
+	constructor(data: object);
+
+	readonly age: number;
+	author: SongAuthor;
+	comments: Comment[];
+	commentsCount: number;
+	description: string;
+	duration: number;
+	embed: Embed;
+	embedURL: string;
+	genre: string;
+	id: string;
+	likes: number;
+	playCount: number;
+	publishedAt: Date;
+	readonly publishedTimestamp: number;
+	streams: SongStreams;
+	streamURL: string;
+	thumbnail: string;
+	title: string;
+	trackURL: string;
+	url: string;
+
+	downloadHLS(options?: m3u8Options): Promise<m3u8Stream>;
+	downloadProgressive(options?: RequestOptions): Promise<IncomingMessage>;
+	toJSON(): SongData;
+	toString(): string;
+}
+
+export namespace Util {
+	function fetchSongStreamURL(songURL: string, clientID: string): Promise<string>;
+	function keygen(force?: boolean): Promise<string>;
+	function last(arr: any[]): void;
```

Comment:
Hello, this part is incorrect https://github.com/DevSnowflake/soundcloud-scraper/blob/master/src/util/Util.js#L19

Reply:
I was just taking it raw from the documentation, you might wanna check this out too then if that's the case
![](https://i.newtt.me/r/ksmbgs87t9a.png)

Reply:
I have pushed a new commit that uses a generic to let people specify the type of their input array so that the output is also typed - does that solve your problem?